neuralmonkey/encoders/cnn_encoder.py

```diff
@@ -1,18 +1,26 @@
 """CNN for image processing."""
 
-from typing import List, Tuple, Optional
+from typing import cast, List, Tuple, Optional, Set, Union
+from typeguard import check_argument_types
 
 import numpy as np
 import tensorflow as tf
 
-from tensorflow.contrib.layers import conv2d, max_pool2d
-
 from neuralmonkey.checking import assert_shape
 from neuralmonkey.dataset import Dataset
 from neuralmonkey.decorators import tensor
 from neuralmonkey.model.model_part import ModelPart, FeedDict
-from neuralmonkey.model.stateful import SpatialStatefulWithOutput
+from neuralmonkey.model.stateful import (SpatialStatefulWithOutput,
+                                         TemporalStatefulWithOutput)
 from neuralmonkey.nn.projection import multilayer_projection
+from neuralmonkey.nn.ortho_gru_cell import OrthoGRUCell
+
+
+# pylint: disable=invalid-name
+ConvSpec = Tuple[str, int, int, str, int]
+ResNetSpec = Tuple[str, int, int]
+MaxPoolSpec = Tuple[str, int, int, str]
```

Comment:
Tyhle tři řádky přímo vybízej k tomu, abych napsal, že tu chybí docstring.

Reply:
.. a přehodnotit jestli by nešlo použít NamedTuple

Reply:
No asi nešlo, protože se to má původ v konfiguráku a z něj ti vyždycky přijdou jenom tuply.

Reply:
Ale tupl se na namedtupl namaskuje, ne? to by podle mě nemělo vadit - v konfigu bude tupl ale pracovat se s tim bude jako s namedtuplem

Reply:
Tak jsem to právě vyzkoušel a `check_argument_types` to nesežere:

```
type of argument "convolutions"[0] must be one of (ConvSpec, ResNetSpec, MaxPoolSpec); got tuple instead
```

Reply:
to je divný, ne?

Reply:
Neni. Čeká to konkrétní typ a dostane to jeho předka.

Reply:
to by to padalo i všude jinde ne? třeba když u `OutputProjection` nebo tak.. Navíc `Tuple` neni předkem `tuple`, ani `ConvSpec` neni předkem `Tuple`.

Reply:
Jo ty jsi zkoušel `NamedTuple`? Aha já to splet myslel jsem že jsme v jiný diskusi.